presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java

```diff
@@ -0,0 +1,288 @@
+
+        Optional<OperatorType> optOperatorType = functionAndTypeManager.getFunctionMetadata(conjunct.getFunctionHandle()).getOperatorType();
+        checkArgument(optOperatorType.isPresent() && optOperatorType.get().isComparisonOperator(), "optOperatorType has to be present and has to be a comparison operator");
+
+        OperatorType operatorType = optOperatorType.get();
+        checkArgument(operatorType.equals(LESS_THAN) || operatorType.equals(LESS_THAN_OR_EQUAL), "operatorType has to be < or <=");
+    }
+
+    // If the join is an outer join, we need to make sure any inferred predicate is on the inner side only
+    // (i.e. does not reference any 'outerVariables').
+    //private Optional<RowExpression> compareAndExtractInequalities(int firstConjunctIndex, int secondConjunctIndex, ImmutableList<RowExpression> inequalities)
+    private Optional<RowExpression> compareAndExtractInequalities(RowExpression expression1, RowExpression expression2)
+    {
+        CallExpression firstConjunct = (CallExpression) expression1;
+        OperatorType firstOperatorType = functionAndTypeManager.getFunctionMetadata(firstConjunct.getFunctionHandle()).getOperatorType().get();
+
+        CallExpression secondConjunct = (CallExpression) expression2;
+        OperatorType secondOperatorType = functionAndTypeManager.getFunctionMetadata(secondConjunct.getFunctionHandle()).getOperatorType().get();
+
+        // Get the inference operator, or empty if no inference possible.
+        Optional<OperatorType> inferenceOperatorType = getComparisonInferenceOperatorType(secondOperatorType, firstOperatorType);
+        if (!inferenceOperatorType.isPresent()) {
+            return Optional.empty();
+        }
+
+        Optional<RowExpression> inferredFirstArgument = Optional.empty();
+        Optional<RowExpression> inferredSecondArgument = Optional.empty();
+        Set<VariableReferenceExpression> variablesReferencedInInferredPredicate = new HashSet<>();
+
+        // For two inequalities "a < b" and "b < 5" we want to infer "a < 5"
+        // Compare the call expressions' arguments to determine equivalence
+        // Make sure one input has no column/variable references since we only want to infer predicates with constants
+        // LHS : Left hand side
+        // RHS : Right hand side
+        RowExpression firstConjunctLHS = firstConjunct.getArguments().get(0);
+        RowExpression firstConjunctRHS = firstConjunct.getArguments().get(1);
+        RowExpression secondConjunctLHS = secondConjunct.getArguments().get(0);
+        RowExpression secondConjunctRHS = secondConjunct.getArguments().get(1);
+        if (expressionEquivalence.areExpressionsEquivalent(firstConjunctRHS, secondConjunctLHS)) {
+            Set<VariableReferenceExpression> referencedVariables = VariablesExtractor.extractUnique(firstConjunctLHS);
+            if (referencedVariables.isEmpty()) {
```

Comment:
This is a lot of repeated logic for each case and a bit hard to follow.  I would extract a method for getting the variablesReferencedInInferredPredicate.

e.g. something like 
```
if (expressionEquivalence.areExpressionsEquivalent(firstConjunctRHS, secondConjunctLHS)) {
     variablesReferencedInInferredPredicate = getVariablesReferencedInInferredPredicate(firstConjunctLHS, secondConjunctRHS);
      if(!variablesReferencedInInferredPredicate.isEmpty()) {
           inferredFirstArgument = Optional.of(firstConjunctLHS);
           inferredSecondArgument = Optional.of(secondConjunctRHS);
       }
}
else if(expressionEquivalence.areExpressionsEquivalent(firstConjunctRHS, secondConjunctLHS)) {
     variablesReferencedInInferredPredicate = getVariablesReferencedInInferredPredicate(firstConjunctRHS, secondConjunctLHS);
      if(!variablesReferencedInInferredPredicate.isEmpty()) {
           inferredFirstArgument = Optional.of(firstConjunctRHS);
           inferredSecondArgument = Optional.of(secondConjunctLHS);
       }
}
```
 and then
```
private static Set<VariableReferenceExpression> getVariablesReferencedInInferredPredicate(RowExpression firstConjunct, RowExpression secondConjunct)
{
    Set<VariableReferenceExpression> firstConjunctReferencedVariables = VariablesExtractor.extractUnique(firstConjunct);
     if(firstConjunctReferencedVariables.isEmpty()) {
         return VariablesExtractor.extractUnique(secondConjunct);
     } 
    Set<VariableReferenceExpression> secondConjunctReferencedVariables 
= VariablesExtractor.extractUnique(secondConjunct);
    if(secondConjunctReferencedVariables.isEmpty()) 
    { 
       return firstConjunctReferencedVariables;
    }
   // the inferred predicate cannot reference variables from both expressions
  // return empty to indicate that there is no valid inferred predicate
   return ImmutableSet.of();
 }
```

Reply:
Thanks! This is easier to parse